More arm fixes

mvwieringen opened the pull request:
First 3 are simple ones. The bdiff one is changing to use feof() which is essentially the same as the current code. From then
on its gets more and more hairy. Format was doable but on testing couldn't select the disk but after some more hacking got
it to a state that you can do format, select disk, partition, print.

The rest is mostly for document formating with troff/nroff etc. That code is really somewhat nastier and may need some more
eyes.

mvwieringen commented:
I added some more network stuff and krb5, this is probably as far as we can get with simple changes.

mvwieringen commented:
Did a test nroff -man run on a manual page and it seems I haven't fully bricked nroff/eqn/tbl (not sure all are used).

richlowe commented on usr/src/cmd/csplit/csplit.c:
```diff
@@ -369,7 +369,7 @@ getfile()
 		if (strlen(fptr) > fiwidth) {
 			fatal("Suffix longer than %ld chars; increase -n\n",
-			    (char *)fiwidth);
+			    (char *)(uintptr_t)fiwidth);
```
shouldn't this just be `fiwidth`?  What even is the old code doing?

mvwieringen replied:
Its passing an int value via a char * pointer and using the %ld pattern in fatal to print the value of the integer. 
Yes I know dirty as hell. But I guess this is the "cleanest" to get it working.

mvwieringen replied:
Ok wrote a cleaner workaround.

richlowe commented on usr/src/cmd/format/io.c:
```diff
@@ -1002,7 +1002,7 @@ input(int type, char *promptstr, int delim, u_ioparam_t *param, int *deflt,
 		 * alloc a copy of the string and return it
 		 */
-		return ((int)alloc_string(token));
+		return ((intptr_t)alloc_string(token));
```
This function returns a `uint64_t`, though it seems like it really should return a `uintptr_t`.  Better to fix the function.
I have no idea why it returns an integer at all though, what the heck is this doing?

mvwieringen replied:
If I look at misc.c which implements alloc_string its char *alloc_string(char *s)

mvwieringen replied:
I tried a couple of times to come up with something better but honestly this is the "cleanest".

richlowe replied:
This is my fault.  By "this function" I meant `input` not `alloc_string`.  I think the cast in the `return` should match the signature of the function we're returning from.  Not that of `alloc_string`.  Sorry.

mvwieringen replied:
Done

richlowe commented on usr/src/cmd/tput/tput.c:
```diff
@@ -232,7 +232,7 @@ outputcap(char *cap, int argc, char **argv)
 				parm[parmset] = atoi(argv[optind]);
 			else
-				parm[parmset] = (int)argv[optind];
+				parm[parmset] = (int)(uintptr_t)argv[optind];
```
in libcurses these are longs, so probably casting them to int was always a mistake, but now it will truncate the pointer and that's bad.  I think.  I think because `tparm` doesn't seem to be documented _well_ anywhere, but I get the impression that for some parameters we sometimes use this _as a pointer_, and so the truncation will crash.  But it's only an impression.  Wow curses is bad.

I think you want to cast these to long, and maybe use strtol above?

mvwieringen replied:
Yeah I'm pretty sure it doesn't work at all. I will try the strtol() as then it makes some sense.

mvwieringen replied:
Or even an atol(). Not sure what to do here.

richlowe commented on usr/src/cmd/troff/n3.c:
```diff
@@ -695,7 +695,7 @@ int	x;
 		edone(0100);
 	}
-	if (j = (unsigned)i % sizeof(int)) {	/*check alignment for 3B*/
+	if (j = (unsigned long)i % sizeof(int)) {	/*check alignment for 3B*/
```
We _probably_ don't need this code in general.  We're definitely not on a 3B!
If we keep this code, the `sizeof` should be a long too, or it should use `IS_P2ALIGNED` or something.

mvwieringen replied:
Done

richlowe commented on usr/src/cmd/cmd-inet/common/kcmd.c:
```diff
@@ -751,7 +751,7 @@ deswrite_compat(int fd, char *buf, int len, int secondary)
 	if (krb5_c_encrypt_length(kcmd_context, final_enctype,
 			use_ivecs ? (size_t)len + 4 : (size_t)len, &ret)) {
-		desoutbuf.length = ((size_t)-1);
+		desoutbuf.length = ((ssize_t)-1);
```
`desoutbuf.length` is an `int`, should that change too?

The second worry is perhaps more important, is what they _thought_ would happen that this would make `desoutbuf.length` positive so it tripped the check at line 760?  'cos that won't happen either before or after your change.

danmcd replied:
desoutbuf.length is an `unsigned int` per `usr/src/uts/common/gssapi/mechs/krb5/include/krb5.h`'s definition of krb5_data.

richlowe replied:
I merged this (without casting -1), thanks Dan.

richlowe commented on usr/src/cmd/rcm_daemon/common/rcm_event.c:
```diff
@@ -121,8 +121,8 @@ process_event(int cmd, int seq_num, nvlist_t *nvl, nvlist_t **ret)
 	uint64_t pid64 = (uint64_t)0;
-	size_t buflen = 0;
-	size_t interval_size = 0;
+	uint_t buflen = 0;
```
It feels like these being `size_t` was correct, and whatever used them was probably wrong, but I'm not sure.

mvwieringen replied:
That would mean changing nvlist_lookup_byte_array() the fourth argument.

richlowe replied:
Drat.  Ok, then yeah, this makes sense.

richlowe commented on usr/src/cmd/rcm_daemon/common/rcm_impl.h:
```diff
@@ -103,7 +103,7 @@ typedef struct rcm_queue {
 #define	RCM_STRUCT_BASE_ADDR(struct_type, x, y)		\
 	((struct_type *) ((void *)(((char *)(x)) -	\
-			(int)(&((struct_type *)0)->y))))
+			(intptr_t)(&((struct_type *)0)->y))))
```
This makes my brain hurt.

The second line, where you added the cast, is `offsetof` which would be better to just use.  The first line _might_ make this an equivalent to `containerof` (see `sys/containerof.h`.

Using `offsetof` is definitely better, if the whole things devolves into `containerof` just using `containerof` is better, but I'm not certain it does.

mvwieringen replied:
The second part indeed translates into offsetof(struct_type, y)))) reading the definition of containerof() indeed is the
same when you replace the second part with offsetof().

#define RCM_STRUCT_BASE_ADDR(struct_type, x, y) \
__containerof(x, struct_type, y)

richlowe commented on usr/src/cmd/Makefile:
```diff
@@ -346,6 +346,7 @@ COMMON_SUBDIRS=	\
 	setmnt		\
 	setpgrp		\
 	sgs		\
+	sh		\
```
I wasn't going to bother, at least as a first pass, shipping the legacy `usr/has` stuff on ARM.

mvwieringen replied:
Ok I drop this then,

richlowe commented on usr/src/cmd/cmd-inet/usr.sbin/Makefile:
```diff
@@ -66,12 +66,12 @@ SRCS=		$(PROGSRCS) $(OTHERSRC)
 SUBDIRS=	in.rdisc \
 		in.talkd inetadm inetconv ipmpstat ipqosconf \
-		nwamadm nwamcfg routeadm \
+		nwamadm nwamcfg routeadm snoop \
```
I'd like to ask @jclulow and @citrus-it to take a look at this, because snoop has been such a hot bed of security problems where it reads untrusted data as root so much.

mvwieringen replied:
I redid the snoop patch a bit its now cleaner then the previous one mostly struct timeval to struct timeval32 and some
fallout from that as localtime() needs a time_t so we need an extra local variable to store the tv_sec in from the
struct timeval32 tv_sec member. Rest is the same the ntp id cast is the same a most other casts for ntp stuff seems
that one was just forgotten or was not a problem before but is now. Then there is the missing header which is also
fixed.

richlowe reviewed (commented):
I haven't looked at roff, etc, because the implementation there (prior to your changes) makes my head spin.  I've picked up most other bits of this I've not commented on to push in a little while. Thanks!

richlowe commented on usr/src/pkg/manifests/service-security-kerberos-5.p5m:
```diff
@@ -29,32 +29,29 @@ set name=pkg.description value="Kerberos version 5 support"
-$(not_aarch64)dir path=etc/krb5 group=sys
-$(not_aarch64)file path=etc/krb5/krb5.conf group=sys \
+dir  path=etc/krb5 group=sys
+file path=etc/krb5/krb5.conf group=sys \
```
This (and some other bits) come out of cmd/gss, so this breaks packaging, etc.

mvwieringen replied:
Think this is fixed now. Did the changes to the manifest again.

richlowe commented on usr/src/cmd/bc/bc.y:
```diff
@@ -764,7 +764,7 @@ routput(int *p)
 		/* part of a bundle */
 		while (*p != 0)
-			routput((int *)*p++);
+			routput((int *)(uintptr_t)*p++);
```
The contents of `p` here come from `b_space` doesn't that still leave this truncated, just the compiler not warning us?  We also, in `bundle()` seem to pull things that are pointers out of a `va_list` as if they were integers, and generally bundling seems like it must still be broken here.  Are we just lucking out because these tend to be in the heap, and thus _usually_ fit in 32bits, or am I missing something?

mvwieringen commented:
> I haven't looked at roff, etc, because the implementation there (prior to your changes) makes my head spin. I've picked up most other bits of this I've not commented on to push in a little while. Thanks!

Yeah the roff stuff is dirty as hell I got tears in my eyes. Its all of questionable quality by todays standards but I tried to make
the majority work but it might fall over pretty bad. That I could format a manpage was some reassurance it was not completely
broken.

mvwieringen commented:
Ok, reordered the branch a bit. First ones until eqn are the most questionable ones. The rest from csplit on should now be in good enough shape for final review. Not that its all great code but reasonable for now until we start refactoring some stuff big time. The
gss code has lots of (uint_t)sizeof constructs which are mostly useless too but left them for now.

mvwieringen committed:
build 64 bit bc

mvwieringen committed:
build 64 bit rcm_daemon

mvwieringen committed:
build 64 bit csplit

mvwieringen committed:
build 64 bit troff/nroff

mvwieringen committed:
build 64 bit tput

mvwieringen committed:
build 64 bit tbl

mvwieringen committed:
build 64 bit refer

mvwieringen committed:
build 64 bit eqn

mvwieringen committed:
build 64 bit format

mvwieringen committed:
build 64 bit krb5

mvwieringen committed:
build 64 bit srchtxt

mvwieringen committed:
build 64 bit truss

Missing some important stuff but this links and the binary mostly works.
Borrowed some code from other platforms (sparc/intel) which may not make
sense at all. The get_arguments() function gets the function arguments
from some registers on sparc and intel but given saveargs is not yet
implemented on arm for now it will return 0 arguments.

mvwieringen committed:
build 64 bit file

mvwieringen committed:
build 64 bit sppptun

mvwieringen committed:
build 64 bit pppd

mvwieringen committed:
Fix ordering.

Restore most of the ordering to be the same as master.

mvwieringen committed:
build 64 bit nohup

mvwieringen committed:
build 64 bit ypcmd

mvwieringen committed:
build 64 bit rcp rlogin rsh

mvwieringen committed:
build 64 bit gss

mvwieringen committed:
build 64 bit in.routed

mvwieringen committed:
build 64 bit snoop

struct timeval --> struct timeval32

richlowe commented:
Hi, I'm not sure why this got closed and deleted?  I was told @citrus-it was reviewing this, and I should look when he was done?

mvwieringen commented:
I went for an other approach and as such deleted this pull request. Two patches will come in when things are merged upstream e.g. in.routed and file. And on top of that I cherry picked the ones that are less controversial ones e.g. drop for now the troff stuff given I'm not so sure its of much use.

mvwieringen commented:
Ok @richlowe things have landed upstream so if we can merge master into arm64-gate here I can start upstreaming in smaller steps. @citrus-it was half way reviewing as far as I understand. Hopefully doing smaller upstreams makes the integration somewhat easier. While doing the new break-up found some problems in packaging but as of this morning got everything compiling again and could create a disk image.

richlowe commented:
Ok, I normally do the merge on monday, so that's actually happening now.
